postgres-shared/src/types/geo.rs

```diff
@@ -0,0 +1,104 @@
+extern crate geo;
+
+use postgres_protocol::types;
+use self::geo::{Bbox, LineString, Point};
+use std::error::Error;
+
+use types::{FromSql, ToSql, IsNull, Type};
+
+impl FromSql for Point<f64> {
+    fn from_sql(_: &Type, raw: &[u8]) -> Result<Self, Box<Error + Sync + Send>> {
+        if raw.len() != 16 {
+            return Err("invalid message length".into());
+        }
+
+        let x = types::float8_from_sql(&raw[0..8])?;
+        let y = types::float8_from_sql(&raw[8..16])?;
+        Ok(Point::new(x, y))
+    }
+
+    accepts!(Type::Point);
+}
+
+impl ToSql for Point<f64> {
+    fn to_sql(&self, _: &Type, out: &mut Vec<u8>) -> Result<IsNull, Box<Error + Sync + Send>> {
+        types::float8_to_sql(self.x(), out);
+        types::float8_to_sql(self.y(), out);
+        Ok(IsNull::No)
+    }
+
+    accepts!(Type::Point);
+    to_sql_checked!();
+}
+
+impl FromSql for Bbox<f64> {
+    fn from_sql(_: &Type, raw: &[u8]) -> Result<Self, Box<Error + Sync + Send>> {
+        if raw.len() != 32 {
+            return Err("invalid message length".into());
+        }
+
+        let xmax = types::float8_from_sql(&raw[0..8])?;
+        let ymax = types::float8_from_sql(&raw[8..16])?;
+        let xmin = types::float8_from_sql(&raw[16..24])?;
+        let ymin = types::float8_from_sql(&raw[24..32])?;
+        Ok(Bbox{xmax: xmax, ymax: ymax, xmin: xmin, ymin: ymin})
+    }
+
+    accepts!(Type::Box);
+}
+
+impl ToSql for Bbox<f64> {
+    fn to_sql(&self, _: &Type, out: &mut Vec<u8>) -> Result<IsNull, Box<Error + Sync + Send>> {
+        types::float8_to_sql(self.xmax, out);
+        types::float8_to_sql(self.ymax, out);
+        types::float8_to_sql(self.xmin, out);
+        types::float8_to_sql(self.ymin, out);
+        Ok(IsNull::No)
+    }
+
+    accepts!(Type::Box);
+    to_sql_checked!();
+}
+
+impl FromSql for LineString<f64> {
+    fn from_sql(_: &Type, raw: &[u8]) -> Result<Self, Box<Error + Sync + Send>> {
+        if raw.len() < 5 {
+            return Err("invalid message length".into());
+        }
+
+        // let _ = types::bool_from_sql(&raw[0..1])?; // is path open or closed
```

Comment:
This will not cause a database error to occur.

The more interesting part of the issue is with ToSql where there are two different ways to create a path:  an open path `path '[(1,2),(3,4)]'` and a closed path `path '((1,2),(3,4))'`.
The PR handles this by always inserting an open path.  To insert a closed path, you would have to write `pclose($1)` and let Postgres flip that bit.   This behavior is documented in the README.

I opted for that behavior because:
 * The Polygon type is already very similar to a Closed path (see https://www.postgresql.org/docs/9.2/static/datatype-geometric.html#DATATYPE-GEO-TABLE)
 * A LineString is treated like an Open path in most uses